OpenRA.Mods.Common/Traits/CreatesShroud.cs

```diff
@@ -20,8 +20,19 @@ public class CreatesShroud : AffectsShroud
 	{
 		public CreatesShroud(Actor self, CreatesShroudInfo info)
 			: base(self, info) { }
-		protected override void AddCellsToPlayerShroud(Actor self, Player p, PPos[] uv) { p.Shroud.AddProjectedShroudGeneration(self, uv); }
-		protected override void RemoveCellsFromPlayerShroud(Actor self, Player p) { p.Shroud.RemoveShroudGeneration(self); }
-		protected override bool IsDisabled(Actor self) { return self.IsDisabled(); }
+
+		protected override void AddCellsToPlayerShroud(Actor self, Player p, PPos[] uv)
+		{
+			if (!IsDisabled(self))
```

Comment:
Declaring `UpgradableTrait` on the parent `AffectsShroud` but then implementing all the bits that make it work in the subclasses is a bit dirty.  Can you please move the `IsTraitDisabled` check up into `AffectsShroud.Tick` instead?